### PR TITLE
Processor and Checkout refactoring

### PR DESCRIPTION
**Pull request by aldesantis**
This is a pretty extensive refactoring of two critical service objects in the extension: `Processor`, which creates and acts on installments, and `Checkout`, which creates orders to fulfill installments.

I have streamlined and simplified both, because the code was very hard to follow and the chance of introducing subtle bugs when touching it was quite high.

A bunch of other APIs and service objects have also been cleaned up and streamlined in the process (e.g., dispatchers, line item builder).

### CODE REVIEW

**Review by igorbp — COMMENTED** [review]
That's great work, @aldesantis! 👏 🎉 

It took me a while to review it. I wasn't familiar with the old code. What I can say is that the new version is much easier to follow.

File: spec/jobs/solidus_subscriptions/process_installment_job_spec.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+RSpec.describe SolidusSubscriptions::ProcessInstallmentJob do
+  it 'creates an order for the installment' do
```
[the remark this answers is not in the document]
Comment (aldesantis) [reply spec/jobs/solidus_subscriptions/process_installment_job_spec.rb:2]: You're right, I think this used to be an integration test and then I stubbed out the checkout service object. I'll fix the description!

**Comment by aldesantis** [comment]
@igorbp should be good for another review!

**Review by igorbp — APPROVED** [review]
Thanks, @aldesantis! LGTM

There are some tests breaking but I don't think it's related 🤔
